GH-41810: [C++] Support cast kernel from (dense or sparse) union to (large) string

llama90 opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change

Support `cast` kernel from (`dense` or `sparse`) union to (large) string

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

### What changes are included in this PR?

- Support `cast` kernel
  - from `dense union` to (large) `string`
  - from `sparse union` to (large) `string` 

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

### Are these changes tested?

Yes. It is passed by existing test cases.

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

### Are there any user-facing changes?

No.

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->

ZhangHuiGui commented on cpp/src/arrow/compute/kernels/scalar_cast_string.cc:
```diff
@@ -510,6 +510,60 @@ void AddBinaryToFixedSizeBinaryCast(CastFunction* func) {
+    BuilderType builder(input.type->GetSharedPtr(), ctx->memory_pool());
+    RETURN_NOT_OK(builder.Reserve(input.length));
+
+    for (int64_t i = 0; i < input.length; ++i) {
```
Maybe it's a better way to implement this by expand `StringFormatter`?
In this way, we can unify it with other type's implementations and shield the logic of converting strings in the current file.

llama90 replied:
Thank you for the review and the insightful suggestions.

That makes sense, and I'll consider applying that.

llama90 replied:
Also, this issue is related: https://github.com/apache/arrow/pull/41831.

felipecrv replied:
> In this way, we can unify it with other type's implementations and shield the logic of converting strings in the current file.

True, but that can prevent optimizations in the future. The approach of taking a scalar function and turning it into an array function by mapping —`array::map(scalar_function: scalar -> scalar) -> array` — is appealing but prevents vectorization techniques.

felipecrv replied:
A string formatter for unions should allocate a vector of string formatters that can do virtual dispatching (and deal with nesting themselves as well). `StringFormatter<T>` performs static dispatch which allows loop-specialization for the non-nested types. But for nested types we will need to setup a vector of `VirtualStringFormatter` (which is actually a tree) so that all the "switching on the type" happens at construction time (beginning of the loop) and invocations inside the loop are following the same function pointers from the vtables.

```cpp
// in header
class VirtualStringFormatter {
  virtual ... = 0;
};

Result<std::unique_ptr<VirtualStringFormatter>> MakeFormatter(
  const std::shared_ptr<DataType>& type);

// in an anon namespace of the .cpp
// one sub-class per `Type::type`
class <T>StringFormatter : public VirtualStringFormatter {
}
// you can use templates to cover most cases delegating to StringFormatter<T>

class UnionStringFormatter : ...
```

This hierarchy would be similar to the builder class hierarchy.

llama90 commented:
I have been examining the code with the intention of using `StringFormatter`, as it seems like a reasonable choice. However, currently, `StringFormatter` mostly handles primitive types and takes time to support complex types such as `Union`, `Struct`, and `List-Like` types.

Not only did I review the code, but I also attempted to modify it. These modifications are related to the Legacy code such as `CastImpl`.

For reference, please see:
https://github.com/apache/arrow/blob/ff9921ffa89585be69ae85674bb365d03cb22ba4/cpp/src/arrow/scalar.cc#L1239-L1250

The reason I am addressing these issues is fundamentally to remove the legacy `CastTo`.

In the short term, there may be code that requires improvement, but I would like to approach this with the following steps:

1. Implement Cast Compute Kernel.
    - #41668 
    - #41669 
    - #41809 
    - #41810
2. Remove Legacy Cast.
    - #39182 
3. Extend `StringFormatter` to support additional types (e.g., Union, Struct, List-Like, etc.).
    - New Issues.
4. Enable the use of `StringFormatter` within the Cast Compute Kernel.
    - New Issues.

Even the existing implementation of `CastTo` (to string) does not utilize `StringFormatter` for complex types.

For reference, please see:
https://github.com/apache/arrow/blob/ff9921ffa89585be69ae85674bb365d03cb22ba4/cpp/src/arrow/scalar.cc#L1309-L1326

What do you think of this approach? I believe this is a way to address each issue and improve the code accordingly. Mentioning everyone who has reviewed the cast code.

cc @kou, @felipecrv, @ZhangHuiGui

llama90 commented:
I have created an issue to add support for nested types in StringFormatter. It seems like addressing this first would be beneficial.

- #41864

ZhangHuiGui commented:
I'm ok  to move these issues forward with this way.

felipecrv commented on cpp/src/arrow/compute/kernels/scalar_cast_string.cc:
```diff
@@ -510,6 +510,60 @@ void AddBinaryToFixedSizeBinaryCast(CastFunction* func) {
+      const ArraySpan& child_span = input.child_data[union_type.child_ids()[type_id]];
+
+      std::shared_ptr<Scalar> child_scalar;
+      auto child_index = union_type.mode() == UnionMode::DENSE ? offsets[i] : i;
```
This should be a template parameter (a `UnionMode::type`) so we can specialize with `if constexpr` checks in the loop.

felipecrv commented:
@llama90 what is the motivation for working on these kernels? It's hard to define what would be a desired string representation of nested types (something for humans to read? JSON-like notation?) and it's quite a lot of work.

If you're looking for compute kernel work, may I interest you in making sure run-end encoded [1] types are handled everywhere? Scalar kernels can be fixed generically by applying the transformation to the values and keeping the same run-ends. There might be something like this for handling DICTIONARY automatically as well that you can use for inspiration.

[1] https://arrow.apache.org/docs/format/Columnar.html#run-end-encoded-layout

llama90 commented:
@felipecrv Thank you for your review.

The main reason for handling this issue is to address the following problem:

- #35560

I have implemented `Cast` Kernels for the necessary types. And I wanted to tackle the next issue to finally remove the `Scalar::CastTo` function:

- #39182

However, I encountered a problem (such as in https://github.com/apache/arrow/pull/39192#discussion_r1477056138) where various languages binding C++ produced errors (e.g., Python, Ruby). Recently, I found a hint that implementing the `${TYPES}` to String Cast Kernel could solve the issue while trying to fix the recent problems again.

As you mentioned, the long-term aspect of `"something for humans to read? JSON-like notation?"` is important, but the primary problem I wanted to solve was to remove the `Scalar::CastTo` function while maintaining compatibility with the existing code.

For example, the `Scalar::ToString` function used `CastTo(utf8())` to handle `ToString()`, and I thought I could solve this by implementing the `Cast` Kernel for String Type. This is why I attempted to create these kernels.

https://github.com/apache/arrow/blob/54bece3d4cf28df3e7bd92ed27f62d705a5cac96/cpp/src/arrow/scalar.cc#L978-L996

Thank you again for the in-depth review. I feel a bit overwhelmed about how to proceed. 

Solving new issues is important, but I also want to finish the PRs I have already submitted. It feels like I'm not making much progress. Anyway, I will think more about it based on your review.

llama90 commented:
> Imagine converting an instance of `list<list<struct<utf8, list<union<int64, utf8>>>>>` to string. You need a tree of string formatters prepared before the loop and it has to follow the same shape of the `DataType`. Doing it value-by-value is just as hard, but slower.

I think I understand what you mean. Thank you for your response!

If you proceed with the work on the `StringFormatter` as [mentioned](https://github.com/apache/arrow/pull/41894#pullrequestreview-2111760063), I will review it and continue with the process. Thank you once again.
